### PR TITLE
WorkspaceApiServiceSpec refactoring (no JIRA, though related to work for GAWB-1174)

### PR DESCRIPTION
**Pull request by breilly2**
This is just the refactoring that I had to do before adding a reliable test for a bucketUsage passthrough. Instead of a global mock setup, each test configures the specific mock behavior that it needs. The changes exposed a number of situations where there was mock behavior that looked like it was needed but actually wasn't. I believe these changes bring the existing tests to a state where they pass for the right reasons.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

… the test that it is needed for.

Improved WorkspaceApiServiceSpec to test what should be rather than what should not be.

### CODE REVIEW

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala [hunk 1]
```diff
@@ -27,159 +26,57 @@ class WorkspaceApiServiceSpec extends BaseServiceSpec with WorkspaceApiService w
...
+        .withHeaders(MockUtils.header).withStatusCode(status.intValue))
+  }
 
-    workspaceServer
-      .when(
-        request()
-          .withMethod("POST")
-          .withPath(s"${workspaceBasePath}/${workspace.namespace.get}/${workspace.name.get}/entities/batchUpsert")
-          .withHeader(authHeader))
+  def stubRawlsServiceWithError(method: HttpMethod, path: String, status: StatusCode) = {
```
Comment (breilly2) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:64]: I don't think this is actually called. I meant to remove it before committing.

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala [hunk 2]
```diff
@@ -27,159 +26,57 @@ class WorkspaceApiServiceSpec extends BaseServiceSpec with WorkspaceApiService w
...
-          .withMethod("POST")
-          .withPath(s"${workspaceBasePath}/${workspace.namespace.get}/${workspace.name.get}/entities/batchUpdate")
-          .withHeader(authHeader))
-      .respond(
-        org.mockserver.model.HttpResponse.response()
-          .withHeaders(MockUtils.header)
-          .withStatusCode(NoContent.intValue)
-          .withBody(rawlsErrorReport(NoContent).toJson.compactPrint)
-      )
+  override def beforeEach(): Unit = {
```
Comment (breilly2) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:75]: This is `beforeEach` now instead of `beforeAll` so that stub behaviors don't bleed from one test to the next.

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala [hunk 3]
```diff
@@ -288,397 +185,419 @@ class WorkspaceApiServiceSpec extends BaseServiceSpec with WorkspaceApiService w
   "WorkspaceService Passthrough Tests" - {
 
     "Passthrough tests on the /workspaces path" - {
-      "MethodNotAllowed error is not returned for HTTP GET and POST methods" in {
```
Comment (breilly2) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:291]: POST for this isn't actually a passthrough, so the test has been moved to a new section below.

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala [hunk 4]
```diff
@@ -288,397 +185,419 @@ class WorkspaceApiServiceSpec extends BaseServiceSpec with WorkspaceApiService w
...
+        Patch(updateAttributesPath, "[]") ~> dummyUserIdHeaders("1234") ~> sealRoute(workspaceRoutes) ~> check {
+          status should equal(OK)
         }
       }
     }
 
 
     "Passthrough tests on the /workspaces/%s/%s/acl path" - {
-      "MethodNotAllowed error is not returned for HTTP GET and PATCH methods" in {
-        List(HttpMethods.GET, HttpMethods.PATCH) map {
```
Comment (breilly2) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:337]: PATCH is not actually a passthrough, so the test has been moved to a new section below.

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala [hunk 5]
```diff
@@ -288,397 +185,419 @@ class WorkspaceApiServiceSpec extends BaseServiceSpec with WorkspaceApiService w
...
-            new RequestBuilder(method)(aclPath) ~> dummyUserIdHeaders("1234") ~> sealRoute(workspaceRoutes) ~> check {
-              status shouldNot equal(MethodNotAllowed)
-            }
-        }
-      }
-    }
-
-
-    "Passthrough tests on the /workspaces/%s/%s/clone path" - {
-      "MethodNotAllowed error is not returned for POST method" in {
```
Comment (breilly2) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:348]: POST on /clone is not actually a passthrough, so the test has been moved to a new section below.

**Comment by breilly2** [comment]
Github has made this more difficult to read than it should be. If you have trouble, I recommend checking out the branch and doing a local compare ignoring whitespace.

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala (hunk 2 above)
Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:75]: Please double check this with Travis by re-running the build a few times. I believe we made this change to deal with spurious travis tests specific to this class. The only reliable fix we could find was to use `beforeAll`.

Comment (breilly2) [reply src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:75]: @rushtong Thanks for the suggestion. Is that just a matter of clicking "Restart build" a few times on the build result in Travis or is there a more preferred way?

Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:75]: Exactly. The errors were very spurious - the only way to reproduce was to restart some variable number of times.

**Comment by dvoet** [comment]
👍

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala (hunk 2 above)
Comment (breilly2) [reply src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:75]: I restarted the Travis job a couple dozen times throughout the day yesterday and didn't see any failures. Since there was some known flakiness (GAWB-1297) and I have no evidence that this makes that problem any worse, I think it's worth keeping this clean-up.

Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:75]: 👍

File: src/main/scala/org/broadinstitute/dsde/firecloud/model/WorkspaceACL.scala [hunk 6]
```diff
@@ -75,4 +75,5 @@ object WorkspaceACLJsonSupport {
   implicit val WorkspaceACLFormat = jsonFormat1(WorkspaceACL)
 
   implicit val WorkspaceACLUpdateFormat = jsonFormat2(WorkspaceACLUpdate)
+  implicit val ListOfWorkspaceACLUpdatesFormat = listFormat[WorkspaceACLUpdate]
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/firecloud/model/WorkspaceACL.scala]: why was this needed?

Comment (breilly2) [reply src/main/scala/org/broadinstitute/dsde/firecloud/model/WorkspaceACL.scala]: The test for `PATCH` on `/workspaces/_/_/acl` was previously set up as a passthrough test even though the implementation is not a passthrough. When I split the test out and didn't provide a request body, it returned 400 Bad Request instead of 200 OK. Therefore, I had to provide a `List` of `WorkspaceACLUpdate` which there had not yet been a need to explicitly marshal. Without this line, the test does not compile.

This is the technique I had used in a previous project to configure marshalling of a list of something. However, I don't see any other use of `listFormat` in orchestration or rawls. Is there a different preferred way to accomplish this here?

Comment (MatthewBemis) [reply src/main/scala/org/broadinstitute/dsde/firecloud/model/WorkspaceACL.scala]: if you import `import spray.json.DefaultJsonProtocol._` in WorkspaceApiServiceSpec instead of this, it will work. this has all of the implicits for handling collections (and more). my understanding is that if the application knows how to unmarshal a `List` and it knows how to unmarshal a `WorkspaceACLUpdate`, then it knows how to unmarshal a `List[WorkspaceACLUpdate]`

the reason I bring this up is that this diverges from the Rawls implementation of `WorkspaceACL` (which is where this entire file was pulled from). eventually this code won't exist in orchestration and will be part of a dependency so it can't be different

Comment (breilly2) [reply src/main/scala/org/broadinstitute/dsde/firecloud/model/WorkspaceACL.scala]: Good call. Thanks. Fixed.

**Comment by MatthewBemis** [comment]
:+1:
